solutions/usaco-718.mdx

```diff
@@ -2,29 +2,68 @@
 id: usaco-718
 source: USACO Gold 2017 February
 title: Why Did the Cow Cross The Road II
-author: Brad Ma
+author: Brad Ma, Jesse Choe
 ---
  
 [Official Analysis (C++)](http://www.usaco.org/current/data/sol_nocross_gold_feb17.html)
 
-The recursive approach where you try all configurations of crosswalks has the 
-time complexity $\mathcal{O}(2 ^ N)$, which is extremely slow. We observe that we have 
-many duplicate recursive calls where the configuration has been solved already, so by 
-using a top down dynamic progamming approach with memoization we can significantly reduce 
-the time complexity to $\mathcal{O}(N ^ 2)$. 
+Let's define $\texttt{dp}[i][j]$ as the maximum number of disjoint friendly crosswalks up to field $i$ on the first side of the road and field $j$ on the second side of the road. We can build a crosswalk between fields $i$ and $j$ as long as there isn't a crosswalk between fields $i$, $j - 1$ and $i - 1$, $j$. Thus, $\texttt{dp}[i][j] = \max(\texttt{dp}[i - 1][j], \texttt{dp}[i][j - 1], \texttt{dp}[i - 1][j - 1] + (|\texttt{id}[i] - \texttt{id}[j]| \leq 4))$, where $\texttt{id}[i]$ and $\texttt{id}[j]$ represent the breed IDs of the cows in the $i$-th and $j$-th fields, respectively.
 
-In this approach, we store our answers such that if there ever is a call to it again, 
-the function uses the cached answer instead of recomputing it. A bottom up dynamic 
-progamming approach without recursion can run faster because function calls have a 
-large constant, but it is more difficult to implement and is not necessary to pass all 
-test cases.
+The answer is $\texttt{dp}[N][N]$.
 
 ## Implementation
 
 **Time Complexity:** $\mathcal{O}(N ^ 2)$
 
 <LanguageSection>
-  
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+using ll = long long;
+using vl = vector<ll>;
+
+const ll MAXN = 1000;
+
+ll dp[MAXN + 1][MAXN + 1];
+
+ll max(ll a, ll b, ll c){
```

Comment:
no need for this:

https://github.com/cpinitiative/usaco-guide/pull/2472/files#diff-fc4fd89ff8e2c6ad834ddd8818832f84b36f88df0d003e366e3ee1ad3f3fdc19R96